Importing necessary libraries

In [ ]:
import matplotlib.pyplot as plt
import numpy as np 
import pandas as pd



The first step in exploratory analysis is reading in the data and then exploring 
the variables. It is important to get a sense of how many variables and cases there are, 
the data types of the variables and the range of values they take on.

In [2]:
weather_airport = pd.read_csv('~/Documents/AA_Test/input/weather_burbank_airport.csv') 
charging_session = pd.read_csv('~/Documents/AA_Test/input/charging_sessions.csv')

---------------------------------------------------------------------------------------------

checking the dimensions with shape and the variable data types with dtypes - weather

In [3]:
weather_airport.shape 

(29244, 9)

In [4]:
weather_airport.dtypes 

city                        object
timestamp                   object
temperature                float64
cloud_cover                float64
cloud_cover_description     object
pressure                   float64
windspeed                  float64
precipitation              float64
felt_temperature           float64
dtype: object

checking the dimensions with shape and the variable data types with dtypes - charging 

In [5]:
charging_session.shape 

(66450, 13)

In [6]:
charging_session.dtypes 

Unnamed: 0            int64
id                   object
connectionTime       object
disconnectTime       object
doneChargingTime     object
kWhDelivered        float64
sessionID            object
siteID                int64
spaceID              object
stationID            object
timezone             object
userID              float64
userInputs           object
dtype: object

the output shows us that we are working with a set of 29244 records and 9 columns (weather dataset). Several of the column variables are encoded as numeric data (floats) but a few of them are encoded as "object".

the output shows us that we are working with a set of 66450 records and 13 columns (charging dataset). Several of the column variables are encoded as "object" data (floats) but a few of them are encoded as numeric data (float).

---------------------------------------------------------------------------------------------

Checking the first 15 rows 

In [7]:
weather_airport.head(15)

city            timestamp  temperature  cloud_cover  \
0   Burbank  2018-01-01 08:53:00          9.0         33.0   
1   Burbank  2018-01-01 09:53:00          9.0         33.0   
2   Burbank  2018-01-01 10:53:00          9.0         21.0   
3   Burbank  2018-01-01 11:53:00          9.0         29.0   
4   Burbank  2018-01-01 12:53:00          8.0         33.0   
5   Burbank  2018-01-01 13:53:00          8.0         33.0   
6   Burbank  2018-01-01 14:53:00          7.0         30.0   
7   Burbank  2018-01-01 15:53:00          8.0         34.0   
8   Burbank  2018-01-01 16:53:00         12.0         34.0   
9   Burbank  2018-01-01 17:53:00         16.0         34.0   
10  Burbank  2018-01-01 18:53:00         19.0         34.0   
11  Burbank  2018-01-01 19:53:00         21.0         34.0   
12  Burbank  2018-01-01 20:53:00         22.0         34.0   
13  Burbank  2018-01-01 21:53:00         23.0         34.0   
14  Burbank  2018-01-01 22:53:00         23.0         34.0   

   cloud_cover_description  pressure  windspeed  precipitation  \
0                     Fair    991.75        9.0            0.0   
1                     Fair    992.08        0.0            0.0   
2                     Haze    992.08        0.0            0.0   
3            Partly Cloudy    992.08        0.0            0.0   
4                     Fair    992.08        0.0            0.0   
5                     Fair    992.08        0.0            0.0   
6            Partly Cloudy    992.08        0.0            0.0   
7                     Fair    992.41        0.0            0.0   
8                     Fair    993.39        0.0            0.0   
9                     Fair    994.05        0.0            0.0   
10                    Fair    993.72        0.0            0.0   
11                    Fair    992.41        7.0            0.0   
12                    Fair    991.09        0.0            0.0   
13                    Fair    990.10        6.0            0.0   
14                    Fair    990.43        7.0            0.0   

    felt_temperature  
0                8.0  
1                9.0  
2                9.0  
3                9.0  
4                8.0  
5                8.0  
6                7.0  
7                8.0  
8               12.0  
9               16.0  
10              19.0  
11              21.0  
12              22.0  
13              23.0  
14              23.0

In [8]:
charging_session.head(15)

Unnamed: 0                        id             connectionTime  \
0            0  5e23b149f9af8b5fe4b973cf  2020-01-02 13:08:54+00:00   
1            1  5e23b149f9af8b5fe4b973d0  2020-01-02 13:36:50+00:00   
2            2  5e23b149f9af8b5fe4b973d1  2020-01-02 13:56:35+00:00   
3            3  5e23b149f9af8b5fe4b973d2  2020-01-02 13:59:58+00:00   
4            4  5e23b149f9af8b5fe4b973d3  2020-01-02 14:00:01+00:00   
5            5  5e23b149f9af8b5fe4b973d4  2020-01-02 14:00:13+00:00   
6            6  5e23b149f9af8b5fe4b973d5  2020-01-02 14:09:14+00:00   
7            7  5e23b149f9af8b5fe4b973d6  2020-01-02 14:17:32+00:00   
8            8  5e23b149f9af8b5fe4b973d7  2020-01-02 14:25:38+00:00   
9            9  5e23b149f9af8b5fe4b973d8  2020-01-02 14:27:40+00:00   
10          10  5e23b149f9af8b5fe4b973d9  2020-01-02 14:30:45+00:00   
11          11  5e23b149f9af8b5fe4b973da  2020-01-02 15:04:38+00:00   
12          12  5e23b149f9af8b5fe4b973db  2020-01-02 15:04:38+00:00   
13          13  5e23b149f9af8b5fe4b973dc  2020-01-02 15:15:27+00:00   
14          14  5e23b149f9af8b5fe4b973dd  2020-01-02 15:19:05+00:00   

               disconnectTime           doneChargingTime  kWhDelivered  \
0   2020-01-02 19:11:15+00:00  2020-01-02 17:31:35+00:00        25.016   
1   2020-01-02 22:38:21+00:00  2020-01-02 20:18:05+00:00        33.097   
2   2020-01-03 00:39:22+00:00  2020-01-02 16:35:06+00:00         6.521   
3   2020-01-02 16:38:39+00:00  2020-01-02 15:18:45+00:00         2.355   
4   2020-01-02 22:08:40+00:00  2020-01-02 18:17:30+00:00        13.375   
5   2020-01-02 23:00:41+00:00  2020-01-02 20:13:21+00:00        13.370   
6   2020-01-03 01:57:58+00:00  2020-01-03 00:59:21+00:00        43.477   
7   2020-01-03 00:44:54+00:00  2020-01-02 20:22:40+00:00        32.441   
8   2020-01-02 19:36:05+00:00  2020-01-02 18:38:39+00:00        13.264   
9   2020-01-02 20:01:48+00:00  2020-01-02 19:17:46+00:00        13.285   
10  2020-01-03 00:06:42+00:00  2020-01-02 17:33:07+00:00        11.633   
11  2020-01-02 23:33:26+00:00  2020-01-02 22:48:48+00:00         0.812   
12  2020-01-02 22:08:39+00:00  2020-01-02 22:09:36+00:00        25.567   
13  2020-01-02 23:55:10+00:00  2020-01-02 22:01:53+00:00        39.806   
14  2020-01-02 20:23:34+00:00  2020-01-02 18:12:46+00:00        17.268   

                                 sessionID  siteID  spaceID    stationID  \
0   1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1   1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2   1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   
3   1_1_193_820_2020-01-02 13:59:58.309319       1  AG-1F04  1-1-193-820   
4   1_1_193_819_2020-01-02 14:00:00.779967       1  AG-1F06  1-1-193-819   
5   1_1_194_821_2020-01-02 14:00:05.241573       1  AG-1F13  1-1-194-821   
6   1_1_178_817_2020-01-02 14:09:14.451827       1  AG-1F09  1-1-178-817   
7   1_1_191_804_2020-01-02 14:17:32.329535       1  AG-4F34  1-1-191-804   
8   1_1_194_826_2020-01-02 14:25:37.578692       1  AG-1F11  1-1-194-826   
9   1_1_178_823_2020-01-02 14:27:39.582337       1  AG-1F08  1-1-178-823   
10  1_1_179_788_2020-01-02 14:30:45.474776       1  AG-3F22  1-1-179-788   
11  1_1_194_818_2020-01-02 15:04:37.511153       1  AG-1F14  1-1-194-818   
12  1_1_178_824_2020-01-02 15:04:38.051735       1  AG-1F07  1-1-178-824   
13  1_1_194_822_2020-01-02 15:15:26.800815       1  AG-1F12  1-1-194-822   
14  1_1_179_783_2020-01-02 15:19:05.085113       1  AG-3F29  1-1-179-783   

               timezone  userID  \
0   America/Los_Angeles   194.0   
1   America/Los_Angeles  4275.0   
2   America/Los_Angeles   344.0   
3   America/Los_Angeles  1117.0   
4   America/Los_Angeles   334.0   
5   America/Los_Angeles  3519.0   
6   America/Los_Angeles   933.0   
7   America/Los_Angeles  3737.0   
8   America/Los_Angeles   419.0   
9   America/Los_Angeles   651.0   
10  America/Los_Angeles  2020.0   
11  America/Los_Angeles     NaN   
12  Americ

---------------------------------------------------------------------------------------------

Statistical summary of the variables with describe()

In [9]:
weather_airport.describe()

temperature   cloud_cover      pressure     windspeed  precipitation  \
count  29219.000000  29224.000000  29236.000000  29158.000000   29244.000000   
mean      17.885622     30.107993    986.879400      8.566568       0.059360   
std        6.454604      5.400388      3.607065      6.699889       0.471682   
min        2.000000      4.000000    971.000000      0.000000       0.000000   
25%       13.000000     26.000000    984.500000      6.000000       0.000000   
50%       17.000000     33.000000    986.480000      7.000000       0.000000   
75%       22.000000     34.000000    989.110000     13.000000       0.000000   
max       46.000000     47.000000    999.650000     57.000000      18.540000   

       felt_temperature  
count      29218.000000  
mean          17.731433  
std            6.321100  
min            0.000000  
25%           13.000000  
50%           17.000000  
75%           22.000000  
max           42.000000

In weather_airport, variables like temperature and wind speed show the average values, variability, and range, indicating common weather patterns and potential outliers, such as high wind speeds.

In [10]:
charging_session.describe()

Unnamed: 0  kWhDelivered        siteID        userID
count  66450.000000  66450.000000  66450.000000  49187.000000
mean    4726.102062     11.787916      1.472656   1800.494582
std     3568.025394     10.314789      0.499256   2617.548614
min        0.000000      0.501000      1.000000      1.000000
25%     1845.000000      5.097000      1.000000    431.000000
50%     3960.000000      9.140000      1.000000    818.000000
75%     7027.000000     14.183000      2.000000   1805.000000
max    15291.000000    108.797242      2.000000  19923.000000

In charging_session, the kWhDelivered column shows an average of 11.79 kWh per session with considerable variability, suggesting some high-energy sessions. Additionally, userID is missing for 26% of the sessions, highlighting a significant portion of anonymous or incomplete data.

Data types

In [33]:
charging_session[['id', 'sessionID', 'siteID', 'spaceID', 'stationID', 'timezone', 'userID']] = charging_session[['id', 'sessionID', 'siteID', 'spaceID', 'stationID', 'timezone', 'userID']].astype("string")

In [34]:
# Convert multiple columns to datetime format at once
charging_session[['connectionTime', 'disconnectTime', 'doneChargingTime']] = charging_session[['connectionTime', 'disconnectTime', 'doneChargingTime']].apply(pd.to_datetime, errors='coerce')

What does error =' coerce' do in Python?
If the error argument is passed as coerce , then invalid parsing will be set as NaN . If the error argument is passed as ignore , then invalid parsing will return the input.  

In [35]:
for col in weather_airport:
    print (type(weather_airport[col][1]))

<class 'str'>
<class 'str'>
<class 'numpy.float64'>
<class 'numpy.float64'>
<class 'str'>
<class 'numpy.float64'>
<class 'numpy.float64'>
<class 'numpy.float64'>
<class 'numpy.float64'>


In [32]:
for col in charging_session:
    print (type(charging_session[col][1]))

<class 'numpy.int64'>
<class 'str'>
<class 'pandas._libs.tslibs.timestamps.Timestamp'>
<class 'pandas._libs.tslibs.timestamps.Timestamp'>
<class 'pandas._libs.tslibs.timestamps.Timestamp'>
<class 'numpy.float64'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>


Missing data

Checking for missing data in both datasets by counting null values in each column


In [ ]:
missing_data_c = charging_session.isnull().sum()
print(missing_data_c)

Unnamed: 0              0
id                      0
connectionTime          0
disconnectTime          0
doneChargingTime     4088
kWhDelivered            0
sessionID               0
siteID                  0
spaceID                 0
stationID               0
timezone                0
userID              17263
userInputs          17263
dtype: int64


In the `charging_session` dataset:
The `doneChargingTime` column has 4,088 missing values.
The `userID` and `userInputs` columns each have 17,263 missing values, indicating a significant amount of missing data in user-related fields.
All other columns have no missing values.

In [37]:
missing_data_w = weather_airport.isnull().sum()
print(missing_data_w)

city                        0
timestamp                   0
temperature                25
cloud_cover                20
cloud_cover_description    20
pressure                    8
windspeed                  86
precipitation               0
felt_temperature           26
dtype: int64


In the `weather_airport` dataset:
`temperature`, `cloud_cover`, `cloud_cover_description`, `pressure`, `windspeed`, and `felt_temperature` have a small number of missing values (up to 86 in `windspeed`).
Columns like `city`, `timestamp`, and `precipitation` are fully populated.

Duplicates

Detect and display duplicates

In [ ]:
duplicates_c = charging_session[charging_session.duplicated()]
print("Duplicate rows:")
print(duplicates_c)

Duplicate rows:
Empty DataFrame
Columns: [Unnamed: 0, id, connectionTime, disconnectTime, doneChargingTime, kWhDelivered, sessionID, siteID, spaceID, stationID, timezone, userID, userInputs]
Index: []


Remove duplicates from the DataFrame

In [ ]:
charging_session_cleaned = charging_session.drop_duplicates()
print("Duplicates removed. New shape:", charging_session_cleaned.shape)

Duplicates removed. New shape: (66450, 13)


In [40]:
duplicates_w = weather_airport[weather_airport.duplicated()]
print("Duplicate rows:")
print(duplicates_w)

Duplicate rows:
Empty DataFrame
Columns: [city, timestamp, temperature, cloud_cover, cloud_cover_description, pressure, windspeed, precipitation, felt_temperature]
Index: []


In [41]:
weather_airport_cleaned = weather_airport.drop_duplicates()
print("Duplicates removed. New shape:", weather_airport_cleaned.shape)

Duplicates removed. New shape: (29244, 9)


No duplicate rows present in the datasets